use-cases/using-Intel-QAT-and-kata.md

```diff
@@ -0,0 +1,484 @@
+
+```sh
+$ mkdir -p $QAT_CONF_LOCATION
+$ cp $QAT_SRC/build/c6xxvf_dev0.conf.vm $QAT_CONF_LOCATION/c6xxvf_dev0.conf
+$ sed -i 's/\[SSL\]/\[SHIM\]/g' $QAT_CONF_LOCATION/c6xxvf_dev0.conf
+$ for i in {1..15}; do cp $QAT_CONF_LOCATION/c6xxvf_dev0.conf $QAT_CONF_LOCATION/c6xxvf_dev$i.conf; done
+```
+
+### Expose and Bind Intel QAT virtual functions to VFIO-PCI (Every reboot)
+
+To enable virtual functions, the host OS should have IOMMU groups enabled. In 
+the UEFI Firmware Intel Virtualization Technology for Directed I/O 
+(Intel VT-d) must be enabled. Also, the kernel boot parameter should be 
+`intel_iommu=on` or `intel_iommu=ifgx_off`. The default in Clear Linux currently 
+is `intel_iommu=igfx_off` which should work with the Intel QAT device. The 
+following commands assume an Intel QAT card is installed, IOMMU is on, and 
+VT-d is enabled. The vendor and device ID will be added to the `VFIO-PCI` driver
+so that each exposed virtual function can be bound to the `VFIO-PCI` driver. 
+Once that is done then each virtual function can be passed into a Kata 
+Containers container using the PCIe device passthrough feature. For 
+Kubernetes, the Intel device plugin for Kubernetes handles the binding of 
+the driver but the VF’s still must be enabled.
+
+```sh
+$ sudo modprobe vfio-pci
+$ QAT_PCI_BUS_PF_NUMBERS=$(lspci | grep Quick | cut -d ' ' -f 1)
+$ QAT_PCI_BUS_PF_1=$(echo $QAT_PCI_BUS_PF_NUMBERS | cut -d ' ' -f 1)
+$ echo 16 | sudo tee /sys/bus/pci/devices/0000:$QAT_PCI_BUS_PF_1/sriov_numvfs
+$ QAT_PCI_ID_VF=$(cat /sys/bus/pci/devices/0000:${QAT_PCI_BUS_PF_1}/virtfn0/uevent | grep PCI_ID)
+$ QAT_VENDOR_AND_ID_VF=$(echo ${QAT_PCI_ID_VF/PCI_ID=} | sed 's/:/ /')
+$ echo $QAT_VENDOR_AND_ID_VF | sudo tee --append /sys/bus/pci/drivers/vfio-pci/new_id
+
+# Loop through all the virtual functions and bind to the VFIO driver
+$ for f in /sys/bus/pci/devices/0000:$QAT_PCI_BUS_PF_1/virtfn*
+  do QAT_PCI_BUS_VF=$(basename $(readlink $f))
+   echo $QAT_PCI_BUS_VF | sudo tee --append /sys/bus/pci/drivers/c6xxvf/unbind
+   echo $QAT_PCI_BUS_VF | sudo tee --append /sys/bus/pci/drivers/vfio-pci/bind
+  done
+
+
+```
+
+### Check Intel QAT virtual functions are enabled
+
+If the following command returns empty then the virtual functions are not 
+properly enabled. This command checks the enumerated device IDs for just the 
+virtual functions. Using the Intel QAT as an example, the physical device ID 
+is `37c8` and virtual function device ID is `37c9`. The following command checks 
+if VF's are enabled for any of the currently known Intel QAT device ID's. The
+following `ls` command should show the 16 vf's bound to vfio-pci.
+
+```
+$ for i in 0442 0443 37c9 19e3; do lspci -d 8086:$i; done
+
+# Check if vf's are bound to vfio-pci
+$ ls -la /sys/bus/pci/drivers/vfio-pci
+```
+
+## Prepare Kata Containers
+
+### Download Kata Kernel Source
+
+This follows the instructions from 
+[https://github.com/kata-containers/packaging/tree/master/kernel](https://
```

Comment:
Please don't create links like this whose "name" is the URL. If you need to do this, just specify the link itself:

> This follows the instructions from https://github.com/kata-containers/packaging/tree/master/kernel ...

But it's probably better to make the text into the link name. Something like:

> This section follows the [packaging kernel instructions](https://github.com/kata-containers/packaging/tree/master/kernel) ...

Reply:
Done